File: fsharp/luhn/Luhn.fs
module Luhn

open System

let valid (number: string) =
    let double x =
        match x + x with
        | y when y > 9 -> y - 9
        | y -> y

    let validate n =
        let checksum =
            Seq.rev n
            |> Seq.map (Char.GetNumericValue >> int)
            |> Seq.mapi (fun i x -> if i % 2 = 1 then double x else x)
            |> Seq.sum

        checksum % 10 = 0

    match number.Replace(" ", "") with
    | n when String.length n < 2 -> false
    | n when not (Seq.forall Char.IsDigit n) -> false
    | n -> validate n

File: ruby/luhn/luhn.rb
module Luhn
  def self.valid?(number)
    number = number.gsub(/\s/, "")
    
    return false if number.size <= 1
    return false if number[/\D/]

    number
      .chars
      .reverse
      .each_with_index
      .map { |d, i| i.odd? ? d.to_i * 2 : d.to_i }
      .map { |d| d > 9 ? d - 9 : d }
      .sum % 10 == 0
  end
end
